streams/src/main/java/org/apache/kafka/streams/kstream/Suppressed.java

```diff
@@ -0,0 +1,88 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream;
+
+import org.apache.kafka.streams.kstream.internals.suppress.SuppressedImpl;
+import org.apache.kafka.streams.kstream.internals.suppress.SuppressedImpl.IntermediateSuppressionImpl;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+import java.time.Duration;
+
+@SuppressWarnings("unused") // complaint about the top-level generic parameters, which actually are necessary
+public interface Suppressed<K, V> {
+    interface TimeDefinition<K, V> {
+        long time(ProcessorContext context, K k, V v);
+    }
+
+    interface EagerBufferConfig<K, V> extends BufferConfig<K, V, EagerBufferConfig<K, V>> {
+
+    }
+
+    interface StrictBufferConfig<K, V> extends BufferConfig<K, V, StrictBufferConfig<K, V>> {
+
+    }
+
+    interface BufferConfig<K, V, BC extends BufferConfig<K, V, BC>> {
+        static <K, V> EagerBufferConfig<K, V> boundedByKeys(final long maxKeys) {
+            return new SuppressedImpl.EagerBCImpl<>(maxKeys);
+        }
+
+        BC bufferKeys(final long maxKeys);
+
+        static <K, V> EagerBufferConfig<K, V> boundedBySize(final long maxBytes) {
+            return new SuppressedImpl.EagerBCImpl<>(maxBytes);
+        }
+
+        BC bufferBytes(final long maxBytes);
+
+        static <K, V> StrictBufferConfig<K, V> unbounded() {
+            return new SuppressedImpl.StrictBCImpl<>();
+        }
+
+        StrictBufferConfig<K, V> unboundedBuffer();
+
+        StrictBufferConfig<K, V> spillToDiskWhenFull();
+
+        StrictBufferConfig<K, V> shutDownWhenFull();
+
+        EagerBufferConfig<K, V> emitEarlyWhenFull();
+    }
+
+    interface IntermediateSuppression<K, V> {
+        static <K, V> IntermediateSuppression<K, V> withEmitAfter(final Duration timeToWaitForMoreEvents) {
+            final IntermediateSuppression<K, V> intermediateSuppression = new IntermediateSuppressionImpl<>();
+            return intermediateSuppression.emitAfter(timeToWaitForMoreEvents);
+        }
+
+        IntermediateSuppression<K, V> emitAfter(final Duration timeToWaitForMoreEvents);
+
+        static <K, V> IntermediateSuppression<K, V> withBufferConfig(final BufferConfig<K, V, ? extends BufferConfig> bufferConfig) {
+            final IntermediateSuppression<K, V> intermediateSuppression = new IntermediateSuppressionImpl<>();
+            return intermediateSuppression.bufferConfig(bufferConfig);
+        }
+
+        IntermediateSuppression<K, V> bufferConfig(final BufferConfig<K, V, ? extends BufferConfig> bufferConfig);
+    }
+
+    static <K extends Windowed, V> Suppressed<K, V> emitFinalResultsOnly(final StrictBufferConfig<K, V> bufferConfig) {
```

Comment:
This is where we insist on strict buffering for final results.

Reply:
There are several buffering options available that provide strict buffers:
* unbounded (just keep allocating more until you get an OOME)
* bounded, shut down gracefully when full
* unbounded, using disk instead of memory

I think this provides plenty of options. I'm not sure we want to get into providing an option that says "emit final results only (unless it turns out you need more memory than you allocated, in which case emit both intermediate and final results)".

1, that's pretty confusing.
2, The main justification for "final results mode" is that the destination is some kind of system that doesn't permit updates. If we *claim* to support final-results-only, but send updates instead, we risk causing harder-to-debug downstream failures.

Reply:
Fair enough.  But IMHO I can't envision a case where users will voluntarily let a production system shut down, but again that's just my opinion.

Reply:
The (maybe imaginary) scenario I had in mind was:
* your business logic depends on exact buffering behavior (otherwise you wouldn't be using "final")
* you want to bound how much memory the app uses
* you discover that in practice the app needs more memory than your bound

It seems like in a situation like this, you would like a graceful shutdown with a clear message so that you can reconsider your options.
* maybe you take a look at your metrics and discover that you can decrease the window grace period, thereby relieving memory pressure
* maybe you change the environment so you can allocate more memory to the task

Unless you are using IQ or doing something time-sensitive like high-frequency trading, shutting the app down for short periods of time when it's misbehaving should be acceptable.

Having hard failures like this can actually be more operationally friendly than grey failures like frequent GC or apps stealing memory from each other. But of course, it depends on the situation.

There is a possibility of implementing more complex behaviors, such as pausing some tasks to allow others to flush more work, but I don't think we need to think about that right now.

That was all for context. At the end of the day, if we don't think we need the "graceful shutdown" mode, I'd rather not implement it.